### DISCUSSION TITLE
I'm converting a dockerfile to envd, and I have a few questions for you

### DISCUSSION THREAD

**Discussion by zhuwenxing** [tensorchord/envd#1103] [Q&A]
* Can we pass an env variable in system such as `PATH` into the value of runtime.environ()
    for example `ENV PATH $PATH:/usr/local/go/bin` --> `runtime.environ(env={"PATH":"$PATH:/usr/local/go/bin"})`

*  Before adding some apt repository, we need to download the public signing key, so how does config.apt_source() deal with this issue?

**Comment by gaocegege** [comment]
Thanks for the issue!

> Can we pass an env variable in system such as PATH into the value of runtime.environ()
for example ENV PATH $PATH:/usr/local/go/bin --> runtime.environ(env={"PATH":"$PATH:/usr/local/go/bin"})

Will you use the PATH in build time or runtime?

> Before adding some apt repository, we need to download the public signing key, so how does config.apt_source() deal with this issue?

Could you please show the existing dockerfile commands?

**Comment by zhuwenxing** [comment]
```
FROM ubuntu:20.04

ARG VERSION=v4.24.5
ARG BINARY=yq_linux_amd64

SHELL ["/bin/bash", "-o", "pipefail", "-c"]

RUN apt-get update && apt-get install -y --no-install-recommends wget apt-transport-https gnupg ca-certificates && \
    wget --no-check-certificate -qO- "https://get.helm.sh/helm-v3.8.1-linux-amd64.tar.gz" | tar --strip-components=1 -xz -C /usr/local/bin linux-amd64/helm && \
    wget  --no-check-certificate  -P /tmp https://mirrors.aliyun.com/kubernetes/apt/doc/apt-key.gpg && \
    apt-key add /tmp/apt-key.gpg && \
    sh -c 'echo deb https://mirrors.aliyun.com/kubernetes/apt/ kubernetes-xenial main > /etc/apt/sources.list.d/kubernetes.list' && \
    apt-get update && apt-get install -y --no-install-recommends \
    build-essential kubectl && \
    apt-get remove --purge -y && \
    rm -rf /var/lib/apt/lists/*
RUN apt-get -y update && DEBIAN_FRONTEND=noninteractive apt-get install -y tzdata

RUN apt-get -y install sudo && apt-get install python3.8 -y && \
	sudo apt-get update -y && sudo apt-get install -y python3.8-dbg && apt-get install -y jq \
    && wget https://github.com/mikefarah/yq/releases/download/${VERSION}/${BINARY} -O /usr/bin/yq \ 
    && chmod +x /usr/bin/yq \
    && apt-get install -y curl

COPY requirements.txt /requirements.txt

COPY Dockerfile /Dockerfile

RUN apt-get update && apt-get install -y python3-pip git vim --fix-missing && python3.8 -m pip install -U pip setuptools

RUN python3.8 -m pip install --no-cache-dir -r /requirements.txt

RUN wget https://go.dev/dl/go1.18.4.linux-amd64.tar.gz && tar -C /usr/local -xzf go1.18.4.linux-amd64.tar.gz
ENV PATH $PATH:/usr/local/go/bin

WORKDIR /root
```

**Reply by zhuwenxing** [reply]
For question 1, I want to set up a go 1.8 environment, in dockerfile the command is 
```
RUN wget https://go.dev/dl/go1.18.4.linux-amd64.tar.gz && tar -C /usr/local -xzf go1.18.4.linux-amd64.tar.gz
ENV PATH $PATH:/usr/local/go/bin
```
I can use run(["wget https://go.dev/dl/go1.18.4.linux-amd64.tar.gz && tar -C /usr/local -xzf go1.18.4.linux-amd64.tar.gz"]) to replase step 1, but i don't konw which API  should be used to replace `ENV PATH $PATH:/usr/local/go/bin`.

I tried `runtime.environ(env={"PATH":"$PATH:/usr/local/go/bin"})` and  `runtime.command(commands={"export":"PATH=$PATH:/usr/local/go/bin"})`. They all don't work as expected (when I exec into the container, and exec `go` command, then it returns error `go command not found. I  need to manually exec `export PATH=$PATH:/usr/local/go/bin`, then it will work well)

For question 2, you can refer to the doc here https://kubernetes.io/docs/tasks/tools/install-kubectl-linux/
![image](https://user-images.githubusercontent.com/12268675/198509904-2f6455f4-b0f2-44c0-a2ec-2715584a0b3e.png)
If I want to apt-get install -y kubectl, not only  `Kubernetes apt repository` should be added, but also the` Google Cloud public signing key` should be added.

**Comment by zhuwenxing** [comment]
with this envd config, it can satisfy the requirements of Python environment and k8s tools. But I can still not find the way to set up a go   lang environment.
```
def install_tools():
    run(
        [
            'curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl"',
            "sudo install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl",
        ]
    )
    run(
        [
            'wget --no-check-certificate -qO- "https://get.helm.sh/helm-v3.8.1-linux-amd64.tar.gz" | tar --strip-components=1 -xz -C /usr/local/bin linux-amd64/helm'
        ]
    )
    run([
        "wget https://github.com/mikefarah/yq/releases/download/v4.24.5/yq_linux_amd64 -O /usr/bin/yq",
        "chmod +x /usr/bin/yq"
    ])
    run (
        ['sh -c "$(curl -fsSL https://raw.github.com/robbyrussell/oh-my-zsh/master/tools/install.sh)"']
    )

def build():
    base(os="ubuntu18.04", language="python3.8")
    install.python_packages( 
        requirements = "./requirements.txt"
    )
    install.apt_packages(
        name=[
            "wget",
            "apt-transport-https",
            "gnupg",
            "ca-certificates",
            "python3.8-dbg",
            "jq",
            "curl",
            "build-essential",
            "git",
            "vim"
        ]
    )
    shell("zsh")
    install_tools()

```

**Reply by gaocegege** [reply]
Hi, do you use `docker build` or `docker buildx build` to build the docker image?

**Reply by gaocegege** [reply]
BTW, do you prefer 18.04, or does your environment **require** ubuntu 18.04?

**Reply by zhuwenxing** [reply]
1. Using `docker build`

2. When using `18.04`,  the image size will be much small, a little more than 1GB. However, when the base image is 20.04, the final image size will be more than 2GB.

**Reply by gaocegege** [reply]
Gotcha. Thanks for your reply.

I tried the dockerfile and build.envd locally. And found some performance issues with the remote buildkit image cache. The PRs are (going to) merged:

- #1125 
- #1126 

After the new release this week, the first build should be much faster than before.

By the way, envd runs a docker load to load the image from the buildkit container to the docker daemon. It still takes ~15s (depends on the CPU and disk). I hope it can be fixed in #51

**Reply by gaocegege** [reply]
And as for the PATH issue, I opened https://github.com/tensorchord/envd/issues/1127 to keep track. I am working in the fix.

**Comment by zhuwenxing** [comment]
I have checked the version in the container when using `base(os="ubuntu18.04", language="python3.8")`, and it is still ubuntu20.04 (gcc version 9.4.0 (Ubuntu 9.4.0-1ubuntu1~20.04.1)). So ubuntu20.04 is the lowest version envd supported?

```
➜  ~ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/9/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:hsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 9.4.0-1ubuntu1~20.04.1' --with-bugurl=file:///usr/share/doc/gcc-9/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++,gm2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-9 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none=/build/gcc-9-Av3uEd/gcc-9-9.4.0/debian/tmp-nvptx/usr,hsa --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 9.4.0 (Ubuntu 9.4.0-1ubuntu1~20.04.1)
```

**Reply by gaocegege** [reply]
https://github.com/tensorchord/envd/blob/main/pkg/lang/ir/system.go#L205

Currently, only ubuntu20.04 is supported. But it should be easy to support 18.04 in envd. I opened an issue to keep track.

- https://github.com/tensorchord/envd/issues/1131

**Comment by zhuwenxing** [comment]
For the golang env, I am trying to use the following command to make it, but it seems that it does not take effect.
```
run(
        [
            "wget https://go.dev/dl/go1.18.4.linux-amd64.tar.gz && tar -C /usr/local -xzf go1.18.4.linux-amd64.tar.gz",
            'echo "export PATH=$PATH:/usr/local/go/bin" > .bashrc',
            'echo "export PATH=$PATH:/usr/local/go/bin" > .zshrc'

        ]
    )
```
I have checked the .bashrc and .zshrc files, `export PATH=$PATH:/usr/local/go/bin` this line was not written into it.

**Reply by gaocegege** [reply]
Interesting workaround. You could try this:

```py
def build():
    shell("zsh")
    install_golang_1_18_4()

def install_golang_1_18_4():
    run(
        [
            "wget https://go.dev/dl/go1.18.4.linux-amd64.tar.gz && tar -C /usr/local -xzf go1.18.4.linux-amd64.tar.gz",
        ]
    )
    # Make PATH substitution in another run enable wget cache.
    run([
            'echo "export PATH=$PATH:/usr/local/go/bin" >> /home/envd/.bashrc',
            'echo "export PATH=$PATH:/usr/local/go/bin" >> /home/envd/.zshrc',
    ])
```

**Reply by zhuwenxing** [reply]
Thanks for the advice, and it works for `envd up`,  since the user is envd. But here is a problem,  when using docker build and run then docker exec -it into the container, the default user is root, so when only build image, it should be  `echo "export PATH=$PATH:/usr/local/go/bin" >> /root/.bashrc`.

When using Dockerfile, we can set the user, but I have not find the ref that how to set user in envd image

**Reply by gaocegege** [reply]
If you use UID 1000 to run `envd up`, the UID in envd environment will be 1000. And it will be UID 0 if you run `envd up` with root privilege.

Do you want to specify the UID by yourself?
